### DISCUSSION TITLE
Using x-for inside of x-if may render stale data

### DISCUSSION THREAD

**Discussion by brandonchinn178** [alpinejs/alpine#3900] [7. Bugs (archived)]
Within a `<template x-if="foo">` component, I expect that nothing in that component will recalculate if `foo` is false, but I'm finding this isn't true with `x-for`.

Minimal repro:
```html
<script src="https://cdn.jsdelivr.net/npm/alpinejs@3.13.3/dist/cdn.min.js" defer></script>
<body x-data="{
    phase: 1,
    get something() {
        if (this.phase !== 1) { throw new Error('Phase is not 1!') }
        return true
    },
    get items() {
        return [1]
    },
}">
    <template x-if="phase === 1">
        <template x-for="item in items">
            <p @click="phase = 2" :something="something">Click me!</p>
        </template>
    </template>
</body>
```
Here, I expect that the error in `something` will never happen, because it's inside an `x-if` template that checks the condition. But if I click the message to set `phase = 2`, I get the error. Specifically, I get the error for every item in `items`. The error only happens inside of `x-for`; removing `x-for` will work as expected.

Is this a bug?

**Comment by ekwoka** [comment]
This would be non-destructive.

I have a PR planned to address this.

Basically, the x-if and x-for remove the elements, and let the removal trigger the elements cleanup, but if the changes that caused the x-if/x-for to cleanup also triggered the effects on the element, they will run once again before the element is cleaned up.

It should be a simple fix to have them actively destroy the element immediately.
